### PR TITLE
Support max-allowed Pods for nodes

### PR DESCRIPTION
**Pull request by islinwb**
xref:
- https://github.com/kubernetes-sigs/poseidon/issues/88
- https://github.com/kubernetes-sigs/poseidon/pull/112

### CODE REVIEW

**Comment by islinwb** [comment]
@shivramsrivastava The PR's ready.

File: src/scheduling/flow/cpu_cost_model.cc [hunk 1]
```diff
@@ -315,7 +329,32 @@ FlowGraphNode* CpuCostModel::GatherStats(FlowGraphNode* accumulator,
...
+      VLOG(2) << "Updating PU " << accumulator->resource_id_ << "'s "
+              << "resource stats!";
+      // Get the CPU stats for this PU
+      string label = rd_ptr->friendly_name();
+      uint64_t idx = label.find("PU #");
+      if (idx != string::npos) {
+        string core_id_substr = label.substr(idx + 4, label.size() - idx - 4);
+        uint32_t core_id = strtoul(core_id_substr.c_str(), 0, 10);
+        float available_cpu_cores =
+                latest_stats.cpus_stats(core_id).cpu_capacity() *
```
Comment (m1093782566) [inline src/scheduling/flow/cpu_cost_model.cc]: I feel a bit confused about why we need to touch the CPU stats?

File: src/scheduling/flow/cpu_cost_model.cc [hunk 2]
```diff
@@ -315,7 +329,32 @@ FlowGraphNode* CpuCostModel::GatherStats(FlowGraphNode* accumulator,
...
+                (1.0 - latest_stats.cpus_stats(core_id).cpu_utilization());
+        rd_ptr->mutable_available_resources()->set_cpu_cores(
+                available_cpu_cores);
+      }
+      // Running/idle task count
+      rd_ptr->set_num_running_tasks_below(rd_ptr->current_running_tasks_size());
+      rd_ptr->set_num_slots_below(FLAGS_max_tasks_per_pu);
```
Comment (m1093782566) [inline src/scheduling/flow/cpu_cost_model.cc]: Should not we set the rd.num_slots_below()?

Comment (shivramsrivastava) [reply src/scheduling/flow/cpu_cost_model.cc]: Yes. We need to have alternative way of setting this number to number of allocatable  pods on that particular core(PU). If not core, at least we need to get to know this number for every scheduling round. So that we can remove this constant number. And because of 'FLAGS_max_tasks_per_pu' currently we can't schedule more than 'FLAGS_max_tasks_per_pu' number of pods on that PU or corresponding Machine.

So possible alternative would be,
1) Use very big number which is always greater than allocatable pods for any node.
or
2) Get number of allocatable  pods from poseidon or any stats server like heapster, and use it to replace FLAGS_max_tasks_per_pu.

So in this way, we don't restrict flow from Machine to PU by this number FLAGS_max_tasks_per_pu. So that we can schedule more pods than FLAGS_max_tasks_per_pu for each PU.

I am verifying this part by putting debugs in firmament code, will share more details soon.

File: src/scheduling/flow/cpu_cost_model.cc (hunk 1 above)
Comment (shivramsrivastava) [reply src/scheduling/flow/cpu_cost_model.cc]: Here we are reading latest machine stats sample for a machine. And fetching particular core's cpu utilization, available cpus and storing it in PU's resource descriptor. And then we accumulate combined cpu stats of all PUs(PU - core, machine may more than 1 PUs) at the Machine's resource descriptor. While deciding the cost, drawing arcs we use collected available cpu from machine's resource descriptor.

File: src/scheduling/flow/cpu_cost_model.cc [hunk 3]
```diff
@@ -315,7 +329,32 @@ FlowGraphNode* CpuCostModel::GatherStats(FlowGraphNode* accumulator,
...
+      if (idx != string::npos) {
+        string core_id_substr = label.substr(idx + 4, label.size() - idx - 4);
+        uint32_t core_id = strtoul(core_id_substr.c_str(), 0, 10);
+        float available_cpu_cores =
+                latest_stats.cpus_stats(core_id).cpu_capacity() *
+                (1.0 - latest_stats.cpus_stats(core_id).cpu_utilization());
+        rd_ptr->mutable_available_resources()->set_cpu_cores(
+                available_cpu_cores);
+      }
+      // Running/idle task count
```
Comment (shivramsrivastava) [inline src/scheduling/flow/cpu_cost_model.cc]: 353 to 324 is already done at 322 to 324. So should be removed. When accumulator->type_ is PU both code sections gets executed. Because 'other' is 'SINK' when accumulator is 'PU'.

File: src/scheduling/flow/cpu_cost_model.cc (hunk 2 above)
Comment (shivramsrivastava) [reply src/scheduling/flow/cpu_cost_model.cc]: Current firmament code fixed the maximum number of pods that can be scheduled on PU by using constant number FLAGS_max_tasks_per_pu. So because of this when running_tasks on PU equals FLAGS_max_tasks_per_pu, capacity of arc from Machine to PU is being set to zero. So we are restricted to schedule only FLAGS_max_tasks_per_pu number of tasks/pods on the PU. So in order to remove this restriction we need to set this rd.num_slots_below() to maximum pods that can be scheduled on that PU, that too only for PU.

Code changes can be like.
1)Add new field ‘max_pods’ in ResourceDescriptor, which gets value from kubelet parameter ‘max-pods’ only once when machine is added.
2)For PU node only, while updating the resource descriptor, update num_slots_below to max_pods like below.
rd.set_num_slots_below(machine_rd.max_pods());

In this way, we can schedule max_pods number of pods on that PU not just FLAGS_max_tasks_per_pu.

File: src/scheduling/flow/cpu_cost_model.cc (hunk 3 above)
Comment (islinwb) [reply src/scheduling/flow/cpu_cost_model.cc]: 353 to 354?

File: src/scheduling/flow/cpu_cost_model.cc (hunk 2 above)
Comment (islinwb) [reply src/scheduling/flow/cpu_cost_model.cc]: Currently `max-pods` is passed to `num_slots_below`.  Need a new field?

Comment (shivramsrivastava) [reply src/scheduling/flow/cpu_cost_model.cc]: Since num_slots_below gets changed every scheduling round, This value does not persist. So better to add new field.

File: src/scheduling/flow/cpu_cost_model.cc (hunk 3 above)
Comment (shivramsrivastava) [reply src/scheduling/flow/cpu_cost_model.cc]: Yes. Sorry for the mistake.

File: src/scheduling/flow/cpu_cost_model.cc [hunk 4]
```diff
@@ -898,7 +898,7 @@ FlowGraphNode* CpuCostModel::GatherStats(FlowGraphNode* accumulator,
       }
       // Running/idle task count
       rd_ptr->set_num_running_tasks_below(rd_ptr->current_running_tasks_size());
-      rd_ptr->set_num_slots_below(FLAGS_max_tasks_per_pu);
+      rd_ptr->set_num_slots_below(rd_ptr->max_pods());
```
Comment (islinwb) [inline src/scheduling/flow/cpu_cost_model.cc]: > 2)For PU node only, while updating the resource descriptor, update num_slots_below to max_pods like below.

@shivramsrivastava  We should update `set_num_slots_below` with  machine's `max_pods` here? (Get machine ResourceDescriptor using  `machine_res_id_pus_` and `machine_res_id` ?)

**Comment by islinwb** [comment]
@shivramsrivastava PTAL

**Comment by nrshrivatsan** [comment]
# TL;DR
Schedulers should be cognizant about

1. Workload-Priority
2. Predicates/Prerequisite for Workloads

# Priority 

Instead of a discrete logic of MAX permissible Pods on a Node, would you encourage us having a more condition tree based scheduling?

``` Node ---- HAS ----> Pods ---> have ---> Priority```

Schedulers could honor Pod Priority [ `P1`, `P2`, `P3` ] along-side maximum permissible number of Pods.

```Scheduler ---> Node A ---> Spawn max `P1` Pods ----> Spaw max `P2` Pods```

# Predicate

### Context
It's a good start to have a `Rate-Limiting` function for workload scheduling. 
However workloads in Production grade systems tend to have criteria like

- Affinity/Anti-affinity
- Taint tolerant workloads
- Volume `dependent`
- Fault-domain resilient

### Suggestion



If we abstract the Rate-limit to a be a type of Predicate for scheduling, we could have an elegant predicate-based posidon scheduling.

## Reasons 

- From Business availability stand point, there is a priority for any workload
- MUST have Pods need to be scheduled first, followed by Nice-to-Have pods

@shivramsrivastava @m1093782566  @islinwb - RFC

**Comment by deepak-vij** [comment]
Thanks for your feedback. This is exactly we currently support as part of Poseidon/Firmament scheduler. Node Filtering (Hard Constraints) -> Priority/Scoring (Soft Constraints) -> Taints -> Volume dependency. We are about to incorporate max. pods/node capability to go along with all of the above. If this is what you meant.

**Comment by nrshrivatsan** [comment]
@deepak-vij  thanks for the comment. Could I request for a few reference hyperlinks to the documentation for Hard & Soft constraints?

**Comment by shivramsrivastava** [comment]
@nrshrivatsan 

1) [Taints and Tolerations](
https://docs.google.com/document/d/15gDgtQtEMZbygWOBVK6Lf5I12STLJZe015dWV2X3jEQ/edit?usp=sharing)
2) [Pod Level Affinity/Anti-Affinity](
https://docs.google.com/document/d/1C68E1Thw7i9NGlHFwmnkgWmNSUK3z0oHdH2mLMK9Eyw/edit?usp=sharing)
3) [Node Level Affinity/Anti-Affinity](
https://docs.google.com/document/d/1Qf_-g4FRL-nlHgnQnENw4iTjIE3FLQPXf__N0fxug78/edit?usp=sharing)

4) Volumes dependent workloads: 
     We are currently working on this one, we support only few volumes types and only support pre-bound PVC's and do not support dynamic binding/storage classes. We will publish a document on this soon.

Can you please elaborate on the Rate-Limiting predicate?
Are you suggesting to introduce Rate-Limiting for pods scheduling, as a part of predicate operation?

**Comment by nrshrivatsan** [comment]
@shivramsrivastava Love the links!

_Rate Limiting_

- Set of Pods could satisfy Predicates
- While scheduling these Pods, since resource quotas might be confined, it would be wise if the Scheduler schedules them in a rate-limited fashion
- Value gains of Rate-limiting : determinism in pod-scheduling rate.
